mapie/utils.py

```diff
@@ -1,17 +1,201 @@
-from typing import Tuple, Optional, Union, Iterable, Any, cast
 from inspect import signature
+from typing import (
+    Any,
+    Callable,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+    cast,
+)
 
 import numpy as np
-from sklearn.utils.validation import _check_sample_weight
-from sklearn.base import RegressorMixin, ClassifierMixin
+
+# from scipy.stats.mstats import winsorize
+from sklearn.base import ClassifierMixin, RegressorMixin
+from sklearn.utils import resample
+from sklearn.utils.validation import _check_sample_weight, _num_samples
 
 from ._typing import ArrayLike
 
 
+def phi1D(
+    x: ArrayLike,
+    B: ArrayLike,
+    fun: Callable[[ArrayLike], ArrayLike],
+) -> ArrayLike:
+    """
+    The function phi1D is used by phi2D. It aims at multiplying the vector of
+    predictions by every refitted estimators by a 1-nan matrix specifying, for
+    each training sample, if it has to be taken into account by the aggregating
+    function, before aggregation
+    """
+
+    return fun(x * B)
+
+
+def phi2D(
+    A: ArrayLike,
+    B: ArrayLike,
+    fun: Callable[[ArrayLike], ArrayLike],
+) -> ArrayLike:
+    """
+    The function phi2D is a loop along the testing set. For each sample of the
+    testing set, it call phi1D to multiply the vector of predictions by the
+    refitted estimators by a 1-nan matrix, to compute the aggragted predictions
+    ignoring the nans
+    """
+    return np.apply_along_axis(
+        phi1D,
+        axis=1,
+        arr=A,
+        B=B,
+        fun=fun,
+    )
+
+
+class JackknifeAB:
+    """Generate a sampling method that resample the training set with
+    possible bootstrap. It can replace KFold as cv argument in the MAPIE
+    class
+
+    Parameters
+    ----------
+    agg_function: str,
+        Choose among:
+
+        - "mean"
+        - "median"
+
+    n_resamplings : number of resamplings
+    n_samples: number of samples in each resampling. By default the size
```

Comment:
By default ``None``, the size of the training set.

Reply:
Corrigé